### PR TITLE
Add radial bases, embeddings and native spherical harmonics to mace-torch (ARCH-1)

### PR DESCRIPTION
**Pull request by steffen-wedig**
Fixes #1559 . Review and merged is blocked until pr for #1556 lands: The init methods in embeddings.py require factories from pydantic configs to create  distance transforms and radial bases.

Contains: Plain-torch rebuilds of the radial layer for the v1 stack, pinned to the frozen legacy numbers. Reference implementation of the spherical harmonics.

- mace_torch.nn.radial: BesselBasis, GaussianBasis, ChebyshevBasis, PolynomialCutoff, ZBLBasis, AgnesiTransform, SoftTransform, RadialMLP and the polynomial_envelope function. Legacy's ChebychevBasis and ChebyshevBasisGeneral collapse onto one ChebyshevBasis with an include_constant switch. Every distance input is named edge_lengths.
- mace_torch.nn.embedding: LinearNodeEmbeddingBlock as a bare weight matrix and RadialEmbeddingBlock returning a frozen RadialEmbedding dataclass; blocks take node_atomic_numbers plus edge_index instead of a one-hot and a species table.
- For the besselbasis i added a guard at x = 0 to ensure that his is differentiable, which the current version did not have.
- mace_torch.backends.reference.sph_harm: closed-form real spherical harmonics in e3nn's convention (y polar axis, no Condon-Shortley phase, component normalisation), written as homogeneous recurrences so the zero vector stays finite and every order is differentiable.

Tests: package tests with committed e3nn fp64 literals, algebraic properties and gradcheck/gradgradcheck; tests/parity created with a dtype fixture and live legacy-vs-v1 parity for all bases, transforms, cutoff paths, ZBL, the node embedding and the harmonics to l = 7 including second derivatives. The 17 radial and embedding rows of the feature inventory pin the parity tests.

sphericart was evaluated as potential default backend for the spherical harmonics. Cursory benchmarking on a B200 vs cueq, e3nn, and a compiled spherical harmonic reference implementation showed the compiled reference implementation coming out on top. Should check again if this still holds once the model is compiled end-to-end.

<details>

<summary>spherical harmonic benchmark</summary>

B200 (daisg205/208), float64

| edges | backend | forward (ms) | forward + backward (ms) | double backward (ms) |
| --- | --- | --- | --- | --- |
| 20,000 | reference | 0.43 | 2.04 | 5.85 |
| 20,000 | reference + torch.compile (default, static) | 0.064 | 0.085 | 1.35 |
| 20,000 | reference + torch.compile (default, dynamic) | 0.067 | 0.093 | 1.20 |
| 20,000 | reference + torch.compile (reduce-overhead, static) | 0.063 | 0.081 | 1.52 |
| 20,000 | e3nn | 0.26 | 2.92 | 3.93 |
| 20,000 | sphericart 1.0.9 (default) | 0.044 | 0.29 | fails |
| 20,000 | sphericart 1.0.9 (2nd derivatives) | 0.043 | 0.42 | 1.33 |
| 20,000 | sphericart 2.0.4 | 0.033 | 0.37 | 2.12 |
| 20,000 | cuequivariance (naive) | 3.00 | 27.95 | 29.94 |
| 20,000 | cuequivariance (uniform_1d) | 0.16 | 1.33 | 2.74 |
| 200,000 | reference | 0.44 | 2.04 | 4.54 |
| 200,000 | reference + torch.compile (default, static) | 0.066 | 0.10 | 1.80 |
| 200,000 | reference + torch.compile (default, dynamic) | 0.069 | 0.10 | 0.94 |
| 200,000 | reference + torch.compile (reduce-overhead, static) | 0.080 | 0.12 | 1.22 |
| 200,000 | e3nn | 0.27 | 1.61 | 3.70 |
| 200,000 | sphericart 1.0.9 (default) | 0.088 | 0.96 | fails |
| 200,000 | sphericart 1.0.9 (2nd derivatives) | 0.088 | 0.91 | 1.37 |
| 200,000 | sphericart 2.0.4 | 0.077 | 0.97 | 1.25 |
| 200,000 | cuequivariance (naive) | 3.02 | 27.92 | 29.49 |
| 200,000 | cuequivariance (uniform_1d) | 0.29 | 1.21 | 2.81 |
| 2,000,000 | reference | 1.47 | 4.05 | 8.29 |
| 2,000,000 | reference + torch.compile (default, static) | 0.37 | 0.46 | 1.47 |
| 2,000,000 | reference + torch.compile (default, dynamic) | 0.32 | 0.45 | 1.09 |
| 2,000,000 | reference + torch.compile (reduce-overhead, static) | 0.39 | 0.55 | 1.66 |
| 2,000,000 | e3nn | 1.17 | 3.59 | 6.85 |
| 2,000,000 | sphericart 1.0.9 (default) | 0.61 | 1.87 | fails |
| 2,000,000 | sphericart 1.0.9 (2nd derivatives) | 0.61 | 3.29 | 6.94 |
| 2,000,000 | sphericart 2.0.4 | 0.58 | 2.74 | 9.27 |
| 2,000,000 | cuequivariance (naive) | 4.59 | 28.63 | 40.67 |
| 2,000,000 | cuequivariance (uniform_1d) | 1.70 | 3.50 | 5.74 |

</details>

Potential followup issues: 

- Implement Solid Harmonics as reference aswell.
- I used a pytest parametrization decorator for deciding whether test should run in fp64 or fp32. Maybe we should use something like this more broadly
- Should we ensure via tests, while we are working on the reforge, that all modules are properly compilable with no graph breaks?
- Probably, magnetic mace shouldnt use the spherical harmonics from sphericart by default,  but it should be a general optional backend.
- Apply cutoff in RadialEmbeddingBlock should probably go into the interaction blocks, which are the downstream consumers of the edge features and the cutoff . This would put responsibility on the consumer, and avoid things that exist currently like MagneticScaleShiftMace dropping the cutoff  edge_feats, _ . (current bug)

---

> [!NOTE]
> **Medium Risk**
> These modules sit on the numerical path for every model (radial features, cutoff ordering, SH convention); mistakes would shift energies/forces, though parity tests and gradgradcheck reduce that risk.
> 
> **Overview**
> Introduces the first **plain-torch** building blocks for the v1 stack, ported from legacy with behavior locked by committed literals, package tests, and in-process **legacy-vs-v1 parity** checks.
> 
> **`mace_torch.nn.radial`** adds Bessel/Gaussian/Chebyshev bases, polynomial cutoff, ZBL pair repulsion, Agnesi/soft distance transforms, and `RadialMLP`. Legacy `ChebychevBasis` / `ChebyshevBasisGeneral` merge into one **`ChebyshevBasis`** (`include_constant`). Graph inputs use **`node_atomic_numbers` + `edge_index`** instead of one-hot + species table; **`BesselBasis`** gets an explicit **r = 0** limit so second derivatives stay finite for force training.
> 
> **`mace_torch.nn.embedding`** adds **`LinearNodeEmbeddingBlock`** (weight init matched to legacy e3nn scaling) and **`RadialEmbeddingBlock`**, which always returns **(features, cutoff)** and enforces **cutoff on raw lengths → transform → basis** (and optional **`apply_cutoff`** pre-multiply).
> 
> **`mace_torch.backends.reference.spherical_harmonics`** is a closed-form **e3nn-compatible** real SH reference (no e3nn dependency), with **`E3NN_AXIS_ORDER`** documented for downstream backends.
> 
> **Packaging/docs:** `mace-torch` now depends on **`torch`** and **`ase`** (covalent radii); target layout doc renames the reference SH module to `spherical_harmonics.py`. **`tests/parity/`** compares v1 to frozen legacy for radial/embedding/ZBL/Chebyshev and SH (including derivatives); **`feature_inventory.md`** pins those parity tests for the radial CLI flags.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 215de11a0630a1a099d794e4d71c8dd34019dd34. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/parity/conftest.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+`tests/parity/` is one of the two places allowed to import both stacks. The
+full harness with process-state snapshot and restore is PAR-1's; until it lands
+the tests here compare pure-math blocks whose only global is the default dtype,
+which the fixture below sets and restores.
+"""
+
+import pytest
+import torch
+
```
Comment (aacostadiaz) [inline tests/parity/conftest.py:11]:
```suggestion

pytest.importorskip("mace_torch", reason="the v1 stack is not installed in this job")
```

**Comment by aacostadiaz** [comment]
@cursor review

**Comment by aacostadiaz** [comment]
@cursor review

**Comment by steffen-wedig** [comment]
PR still blocked by merging the config PR, but I think we should take time and get that one right first.

File: packages/mace-torch/src/mace_torch/nn/radial.py [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+        )
+        reduced_distance = edge_lengths / screening_length
+        screening = torch.sum(
+            self.screening_coefficients
+            * torch.exp(-self.screening_exponents * reduced_distance),
+            dim=-1,
+            keepdim=True,
+        )
+        coulomb = (
+            14.3996 * sender_atomic_numbers * receiver_atomic_numbers
```
Comment (aacostadiaz) [inline packages/mace-torch/src/mace_torch/nn/radial.py:342]: This multiplies a python float by integer tensors so the result takes the default dtype

```suggestion
            14.3996 * (sender_atomic_numbers * receiver_atomic_numbers).to(edge_lengths.dtype)
```
